Components/IncomingNotificationManager.php

```diff
@@ -47,27 +50,54 @@ public function __construct(
     }
 
     /**
-     * @param array $notificationItems
-     * @return \Generator
-     * @throws \Doctrine\ORM\ORMException
+     * @param TextNotification[] $textNotifications
+     * @throws ORMException
      * @throws \Doctrine\ORM\OptimisticLockException
      */
-    public function save(array $notificationItems)
+    public function convertNotifications(array $textNotifications)
     {
-        foreach ($notificationItems as $notificationItem) {
+        foreach ($textNotifications as $textNotificationItem) {
             try {
-                if (!empty($notificationItem['NotificationRequestItem'])) {
+                if (!empty($textNotificationItem->getTextNotification())) {
                     $notification = $this->notificationBuilder->fromParams(
-                        $notificationItem['NotificationRequestItem']
+                        json_decode($textNotificationItem->getTextNotification(), true)
                     );
                     $this->entityManager->persist($notification);
                 }
             } catch (InvalidParameterException $exception) {
-                $this->logger->warning($exception->getMessage());
-                yield new NotificationItemFeedback($exception->getMessage(), $notificationItem);
+                $this->logger->warning(
+                    $exception->getMessage() . " " . $textNotificationItem->getTextNotification()
+                );
             } catch (OrderNotFoundException $exception) {
+                $this->logger->warning(
+                    $exception->getMessage() . " " . $textNotificationItem->getTextNotification()
+                );
+            }
+            $this->entityManager->remove($textNotificationItem);
+        }
+        $this->entityManager->flush();
+    }
+
+    /**
+     * @param array $textNotificationItems
+     * @return \Generator
+     * @throws \Doctrine\ORM\ORMException
+     * @throws \Doctrine\ORM\OptimisticLockException
+     */
+    public function saveTextNotification(array $textNotificationItems): \Generator
+    {
+        foreach ($textNotificationItems as $textNotificationItem) {
+            try {
+                if (!empty($textNotificationItem['NotificationRequestItem'])) {
+                    $textNotification = new TextNotification();
+                    $textNotification->setTextNotification(
+                        json_encode($textNotificationItem['NotificationRequestItem'])
+                    );
+                    $this->entityManager->persist($textNotification);
+                }
+            } catch (ORMException $exception) {
                 $this->logger->warning($exception->getMessage());
-                yield new NotificationItemFeedback($exception->getMessage(), $notificationItem);
+                yield new TextNotificationItemFeedback($exception->getMessage(), $textNotificationItem);
```

Comment:
Should this function be void instead?

Reply:
Currently it only yields on an exception and the yielded count is verified: if yielded than saving did not work.

Better would be that it always yield `TextNotificationItemFeedback` so that the Generator functionality is used.
The generator would be handy when decorating this service (as in extracting the logger, ...).

Reply:
Thanks for your input @bortefi! I'll leave as Generator